File: app/src/main/java/com/castelijns/mmdemo/users/UsersPresenter.java
package com.castelijns.mmdemo.users;

import android.widget.TextView;

import com.castelijns.mmdemo.models.User;

import java.util.Collections;
import java.util.List;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class UsersPresenter implements UsersContract.Presenter {

    private UsersContract.View view;
    private UsersRepo repo;

    private Disposable disposable;

    UsersPresenter(UsersContract.View view) {
        this.view = view;
        repo = UsersRepo.getInstance();
    }

    @Override
    public void start() {
        repo.getAllUsers()
                .subscribeOn(Schedulers.io())
                .doOnNext(users -> repo.cacheData(users))
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<User>>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        disposable = d;
                    }

                    @Override
                    public void onNext(List<User> users) {
                        Collections.sort(users);
                        view.showUsers(users);
                        view.showUserCount(users.size());
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public void stop() {
        if (!disposable.isDisposed()) {
            disposable.dispose();
        }
    }

    @Override
    public void onUserClicked(User user, TextView tvUsername, TextView tvEmail) {
        this.view.showUserDetail(user, tvUsername, tvEmail);
    }
}

File: app/src/main/kotlin/com/castelijns/mmdemo/users/UsersPresenter.kt
package com.castelijns.mmdemo.users

import android.widget.TextView
import com.castelijns.mmdemo.app.BasePresenter
import com.castelijns.mmdemo.models.User
import kotlinx.coroutines.experimental.launch

class UsersPresenter(
        private val view: UsersContract.View
) : BasePresenter() {

    private val usersRepo = UsersRepo

    override fun start() {
        view.showLoading()

        launch {
            try {
                loadUsers()
            } catch (e: Exception) {
                view.showError()
            } finally {
                view.hideLoading()
            }
        }
    }

    private suspend fun loadUsers() {
        val allUsers = usersRepo.getAllUsers().await()

        val users = allUsers.sortedBy { it.name }
        usersRepo.cacheData(users)

        view.showUsers(users)
        view.showUserCount(users.size)
    }

    fun onUserClicked(user: User, tvUsername: TextView, tvEmail: TextView) {
        this.view.showUserDetail(user, tvUsername, tvEmail)
    }
}
